parlai/core/pytorch_data_teacher.py

```diff
@@ -426,15 +426,32 @@ def __init__(self, opt):
         self.datapath = build_data(self.opt)
         self.data_gen = self._data_generator()
         self.length_datafile = os.path.join(self.datapath, 'data_length')
+        self.char_index_file = os.path.join(self.datapath, 'char_index')
         self.datafile = os.path.join(self.datapath, 'data')
         self.training = self.datatype.startswith('train')
+        self.ordered = ('ordered' in self.datatype or
+                        ('stream' in self.datatype and not opt.get('shuffle')))
         self._load_lens()
 
     def __getitem__(self, index):
-        while True:
-            idx, ep = next(self.data_gen)
-            if idx == index:
-                return (index, ep)
+        if self.ordered or not self.training:
+            while True:
+                idx, ep = next(self.data_gen)
+                if idx == index:
+                    return (index, ep)
+        else:
+            episode = []
+            current_idx = index
+            episode_done = False
+            with open(self.datafile) as f:
+                while not episode_done:
+                    ex_offset = self.char_index[current_idx]
+                    f.seek(ex_offset)
+                    example = json.loads(f.readline())
+                    episode.append(example)
+                    episode_done = example['episode_done']
+                    current_idx += 1
+            return (index, episode)
```

Comment:
how much slower is this to load a whole epoch? (e.g. with display_data?)

Reply:
Just to illuminate the use case here: this allows for randomized iteration through a large dataset without loading it into memory. This is helpful for me, as I can pre-BPE tokenize/vectorize a very large corpus, without having to hold the entire thing in memory.

Reply:
excellent!